### ISSUE TITLE
GRPO validation takes a long time...

### ISSUE THREAD

**Issue by lianzheng-research** [siyan-zhao/OPSD#12]
Hello, I trained the Qwen3-1.7B model using GRPO and tested it on the AIME 2024 dataset following the run_eval.sh script. I found that testing with the base model was fast, but the version fine-tuned with GRPO was extremely slow. It couldn't even complete predictions for 30 samples within an hour. Is this normal? How long do you usually take to complete a test? Thanks you!

Here is my testing script:
```
#!/bin/bash
BASE_MODEL="/path/to/Qwen3-1.7B"

# after trained, evaluate the performance of the trained model
NCCL_P2P_DISABLE=1 CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python evaluate_math.py \
  --base_model "$BASE_MODEL" \
  --dataset "aime24" \
  --val_n 16 \
  --temperature 1.0 \
  --tensor_parallel_size 8 \
  --checkpoint_dir /path/to/grpo_qwen317b_openthoughts_20260331/checkpoint-7306
wait

**Comment by lianzheng-research** [comment]
Oh thanks. I found that it takes 2 hours to process one batch.
```
Processed prompts:   7%|████▎                                                           | 32/480 [1:51:04<21:22:43, 171.79s/it, est. speed input: 0.69 toks/s, output: 186.83 toks/s]

**Comment by lianzheng-research** [comment]
I'm sorry and get the following results.

```
======================================================================
FINAL RESULTS
======================================================================
Dataset: AIME24
Thinking Mode: ENABLED
Total problems: 30
Solutions per problem: 16
Total solutions: 480

Metrics:
  Pass@16: 0.00% (0/30)
  Average@16: 0.00% (0/480)
  Majority Vote@16: 0.00% (0/30)

Formatting:
  Formatted (boxed) answers: 0/480
  Format rate: 0.00%
======================================================================

**Comment by siyan-zhao** [comment]
Hi, which checkpoint are you evaluating? From the path it looks like step 7306. In our experiments we only evaluated up to 500 steps, and peak performance typically comes within the first ~100 steps across model sizes. 

The core issue is that this reasoning dataset is not well-suited for GRPO with these models. From the wandb logs (frac_reward_zero_std), 70–90% of batches have zero reward standard deviation, meaning all sampled responses for a problem are either all correct or all wrong and have zero gradient, and most of these are all-correct, so GRPO has limited room to learn. Please refer to Figure 3 (rightmost plot) in the paper for this trend. 

We also find 1.7b collapse easier than 4b and 8b. And 8b is more stable than 4b model in this training. 4b model also show degradation on some tasks like aime24/25 during training. Try evaluating earlier checkpoints (e.g., steps 25-100). We used no KL regularization, which may contribute to the collapsing. Increasing batch size might also help. Hope this helps!

**Comment by lianzheng-research** [comment]
It works for me. Thanks a lot. Do you also verify early checkpoints for OPSD?

**Comment by siyan-zhao** [comment]
Yes, you can refer to this evaluation table here: [evaluation table](https://github.com/siyan-zhao/OPSD?tab=readme-ov-file#evaluation-results-across-tasks-on-qwen3-17b). and please see the new version of the paper with updated results: https://arxiv.org/pdf/2601.18734v3. The code has also been updated, see: https://github.com/siyan-zhao/OPSD?tab=readme-ov-file#updates and training steps info in issue https://github.com/siyan-zhao/OPSD/issues/3
